design-proposals/shadow-node.md

```diff
@@ -0,0 +1,157 @@
+  - the #api-calls kubevirt does.
+  - the storage kubevirt takes. 
+  This concern is addressed on the Scalability section.
+- Adding virt-controller as a "proxy" agent, we now depend on its availability. This could cause an increase in heartbeat timeout in corner cases.
+- To a degree, it is still possible to manipulate pods, by causing VMs to migrate from other nodes. 
+
+## Solution details
+In order to easy review and understand what needs to be changed, the first section will describe different usage of Node object in the virt-handler.
+
+### Node writers on virt-handler
+Kubevirt sets ~170 labels on each node. This is set by the following entities
+1. virt-handler [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/cmd/virt-handler/virt-handler.go#L185) the node unschedulable for VMs at the start and the termination (by SIGTERM only) of the virt-handler process to indicate (best-effort) that the node is not ready to handle VMs.
+
+2. virt-handler/heartbeat:
+   * [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) the node unschedulable for VMs when the stop channel closes (i.e. virt-handler exits).
+   * [labels and annotates](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) the node, updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.
+
+3. virt-handler/node labeller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) various information which are mostly static in nature: cpu model, hyperv features, cpu timer, sev/-es, RT... 
+   * This is triggered both by interval (3 min) and on changes of cluster config.
+
+### Node writers on virt-controller
+1. virt-controller/node-controller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-controller/watch/node.go#L318) the node unschedulable for VMs on heartbeat timeout.
+2. virt-controller/nodetopologyupdater [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-controller/watch/topology/nodetopologyupdater.go#L63) the node with supported topology frequencies.
+
+## Shadow node
+
+The proposed solution is to introduce new CRD that will be shadowing a node "ShadowNode". Every entity that now writes to the node, will be writing to ShadowNode instead, and the new controller will have a responsibility to sync allowed information to Node.
+
+## API Examples
+```yaml
+apiVersion: <apiversion>
+kind: ShadowNode
+metadata:
+  name: <Equivalent to existing name>
+  # Labels and Annotations only
```

Comment:
I think this is a great opportunity, to create a real spec for all that info. The reason we use for instance for heartbeat metadata is because we don't have a place on the k8s node object to properly express information. Now we can.

Reply:
Added to the pros list to the CRD approach

Reply:
Yes, having a typed spec with a well defined list of things that we want on the node, looks to be an opportunity here.

Reply:
@rmohr how about if we make it a bit wider. Instead of setting a CRD for this very specific node sync purpose, define a virt-handler CRD, that would hold this `nodeInfo`  (such as the labels and annotations) and also future stuff that could be used in future virt-handler communication with other components...

Reply:
I am fine either way.